### PR TITLE
(S)FTP MaxFetchSize Doc Polishing and DSL

### PR DESCRIPTION
**Pull request by garyrussell**
Also polish programming tips and tricks docs.

### CODE REVIEW

File: src/reference/asciidoc/ftp.adoc [hunk 1]
```diff
@@ -617,6 +617,27 @@ public class FtpJavaApplication {
...
+* `max-messages-per-poll=2` and `max-fetch-size=4`, the adapter will fetch up to 4 files (if available) and emit the first two (if there are at least two); the next two files will be emitted on the next poll.
+* `max-messages-per-poll=2` and `max-fetch-size` not specified, the adapter will fetch all remote files and emit the first two (if there are at least two); the subsequent files will be emitted on subsequent polls (2-at-a-time); when all are consumed, the remote fetch will be attempted again, to pick up any new files.
+
+IMPORTANT: When deploying multiple instances of an application, a small `max-fetch-size` is recommended to avoid one instance "grabbing" all the files and starving other instances.
+
+Another use for `max-fetch-size` is if you want to stop fetching remote files, but continue to process files that have already been fetched.
+Setting the `matchFetchSize` property on the `MessageSource` (programmatically, via JMX, or via a <<control-bus, control bus>>) effectively stops the adapter from fetching more files, but allows the poller to continue to emit messages for files that have previously been fetched.
```
Comment (artembilan) [inline src/reference/asciidoc/ftp.adoc]: typo: s/matchFetchSize/maxFetchSize

File: src/reference/asciidoc/sftp.adoc [hunk 2]
```diff
@@ -662,6 +662,27 @@ public class SftpJavaApplication {
...
+* `max-messages-per-poll=2` and `max-fetch-size=4`, the adapter will fetch up to 4 files (if available) and emit the first two (if there are at least two); the next two files will be emitted on the next poll.
+* `max-messages-per-poll=2` and `max-fetch-size` not specified, the adapter will fetch all remote files and emit the first two (if there are at least two); the subsequent files will be emitted on subsequent polls (2-at-a-time); when all are consumed, the remote fetch will be attempted again, to pick up any new files.
+
+IMPORTANT: When deploying multiple instances of an application, a small `max-fetch-size` is recommended to avoid one instance "grabbing" all the files and starving other instances.
+
+Another use for `max-fetch-size` is if you want to stop fetching remote files, but continue to process files that have already been fetched.
+Setting the `matchFetchSize` property on the `MessageSource` (programmatically, via JMX, or via a <<control-bus, control bus>>) effectively stops the adapter from fetching more files, but allows the poller to continue to emit messages for files that have previously been fetched.
```
Comment (artembilan) [inline src/reference/asciidoc/sftp.adoc]: DITTO

**Review by artembilan — APPROVED** [review]
Only a couple typos.
Will fix on merge.

I wonder if we should dedicate some effort to extract the common remote functionality description outside of the FTP and SFTP chapter to, for example, some section in the file chapter. Mostly same concepts are applied to any other "remote file" protocol - AWS S3, Google Storage, SMB etc.

**Comment by garyrussell** [comment]
Yes; it's quite annoying to have to put the same docs in both places every time.

It would be pretty thankless work, though, I think - finding all the unique bits.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/36ac9f220378c5cc4c05d420727afb2d9f9b47ba.

Yes, I agree about of that docs work.
Maybe we can just combine both of them to a single (S)FTP chapter and just mention unique bits whenever it is necessary ?

**Pull request closed without merging**